### PR TITLE
Refactor filter path, remove LogicalExpression from DB operation level.

### PR DESCRIPTION
**Pull request by Yuqi-Du**
**What this PR does**:

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java [hunk 1]
```diff
@@ -149,7 +171,7 @@ public FilterMatcher<T> compareValues(
     }
   }
 
-  public static final class MatchStrategyCounter {
+  public final class MatchStrategyCounter {
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java:174]: there is still work to do on this class once we have merged: 

* users of the class need to know if they should strict or greedy, that should be a polymorphic function 
* user of the class needs to know if it should remove from the iterator 

https://github.com/stargate/data-api/issues/1412

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java [hunk 2]
```diff
@@ -54,29 +55,50 @@ public Optional<LogicalExpression> apply(T command) {
...
+        command,
+        captureGroups,
+        filter.logicalExpression(),
+        unmatchedCaptures,
+        matchStrategyCounter);
+    // apply strategy to the resolved root captureGroups
+    return matchStrategyCounter.applyStrategy(strategy, captureGroups);
   }
 
   public void captureRecursive(
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java:69]: should be private

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java [hunk 3]
```diff
@@ -123,7 +145,7 @@ protected Capture(Object marker) {
       this.marker = marker;
     }
 
-    public List<FilterOperation<?>> match(ComparisonExpression t) {
+    public List<FilterOperation> match(ComparisonExpression t) {
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java:148]: Raw use of parameterized class 'FilterOperation'

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatchRule.java [hunk 4]
```diff
@@ -17,10 +17,32 @@
  *
  * <p>T - Command type to match
  */
-public record FilterMatchRule<T extends Command & Filterable>(FilterMatcher<T> matcher)
-    implements BiFunction<CommandContext, T, Optional<LogicalExpression>> {
+// public record FilterMatchRule<T extends Command & Filterable>(FilterMatcher<T> matcher)
+//    implements BiFunction<CommandContext, T, Optional<LogicalExpression>> {
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatchRule.java:21]: remove commented code

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterResolver.java [hunk 5]
```diff
@@ -35,10 +35,6 @@ protected FilterResolver(OperationsConfig operationsConfig) {
 
     matchRules =
         Objects.requireNonNull(buildMatchRules(), "buildMatchRules() must return non null");
-    if (matchRules.isEmpty()) {
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterResolver.java:38]: why was this removed ?

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+  /**
+   * Get the {@link CaptureGroup} that has the result of the {@link FilterMatcher.Capture} created
+   * with the supplied marker. If there is no target marker in captureGroup map, return an empty
+   * optional
+   *
+   * @param marker
+   * @return Optional<CaptureGroup>
+   */
+  public Optional<CaptureGroup<?>> getGroupIfPresent(Object marker) {
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:59]: This would be easier to use if it was:

```java
consumeGroupIfPresent(Object marker, Consumer<CaptureExpression<DataType>>>)
```

would simplify the repetative pattern below 
```java
          captureGroups
              .getGroupIfPresent(DYNAMIC_DOCID_GROUP)
              .ifPresent(
                  captureGroup -> {
                    CaptureGroup<Object> dynamicDocIDGroup = (CaptureGroup<Object>) captureGroup;
                    dynamicDocIDGroup.consumeAllCaptures(
```

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java [hunk 7]
```diff
@@ -137,18 +136,19 @@ default Operation resolveDatabaseCommand(CommandContext<DatabaseSchemaObject> ct
    * @param dataApiRequestInfo
    * @param jsonApiMetricsConfig
    * @param command
-   * @param logicalExpression
+   * @param
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java:139]: missing doc

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java [hunk 8]
```diff
@@ -157,24 +157,24 @@ default void addToMetrics(
...
+    getIndexUsageTags(dbFilterLogicalExpression, baseIndexUsage);
     tags = tags.and(baseIndexUsage.getTags());
 
     meterRegistry.counter(jsonApiMetricsConfig.indexUsageCounterMetrics(), tags).increment();
   }
 
-  private void getIndexUsageTags(LogicalExpression logicalExpression, IndexUsage indexUsage) {
-    for (ComparisonExpression comparisonExpression : logicalExpression.comparisonExpressions) {
-      getIndexUsageTags(comparisonExpression, indexUsage);
+  private void getIndexUsageTags(
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java:166]: rename , this does not return a value. recordIndexUsage

can merge the function below into this one.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java [hunk 9]
```diff
@@ -157,24 +157,24 @@ default void addToMetrics(
...
+      DBFilterLogicalExpression dbFilterLogicalExpression, IndexUsage indexUsage) {
+    for (DBFilterBase dbFilter : dbFilterLogicalExpression.getDbFilterList()) {
+      getIndexUsageTags(dbFilter, indexUsage);
     }
-    for (LogicalExpression innerLogicalExpression : logicalExpression.logicalExpressions) {
-      getIndexUsageTags(innerLogicalExpression, indexUsage);
+    for (DBFilterLogicalExpression innerDbFilterLogicalExpression :
+        dbFilterLogicalExpression.getDbFilterLogicalExpressionList()) {
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java:172]: try to use the term sub-expression for these

**Review by amorton — CHANGES_REQUESTED** [review]
some issues with code style and naming, we will go through together

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java [hunk 10]
```diff
@@ -4,9 +4,9 @@
 /**
  * @param value Literal value to use as RHS operand in the query
  * @param type Literal value type of the RHS operand in the query
- * @param <T> Data type of the object
+ * @param <DataType> Data type of the object
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java:7]: Thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/cql/ExpressionUtils.java [hunk 11]
```diff
@@ -52,12 +53,13 @@ public static <K> Or<K> orOf(Expression<K>... expressions) {
   }
 
   public static <K> Expression<K> buildExpression(
-      List<? extends Expression<K>> expressions, String logicOperator) {
+      List<? extends Expression<K>> expressions,
+      DBFilterLogicalExpression.DBLogicalOperator logicOperator) {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/cql/ExpressionUtils.java:57]: resolved, thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/collections/ExpressionBuilder.java [hunk 12]
```diff
@@ -105,58 +105,58 @@ private static Expression<BuiltCondition> buildExpressionRecursive(
...
-        if (dbFilter instanceof AllCollectionFilter allFilter) {
-          List<BuiltCondition> allFilterConditions = allFilter.getAll();
-          List<Variable<BuiltCondition>> allFilterVariables =
-              allFilterConditions.stream().map(Variable::of).toList();
-          conditionExpressions.add(
-              allFilter.isNegation()
-                  ? ExpressionUtils.orOf(allFilterVariables)
-                  : ExpressionUtils.andOf(allFilterVariables));
-        } else if (dbFilter instanceof InCollectionFilter inFilter) {
+    // second for loop, is to iterate dbFilters
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/collections/ExpressionBuilder.java:108]: yes

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
+package io.stargate.sgv2.jsonapi.service.operation.query;
+
+import io.stargate.sgv2.jsonapi.api.model.command.clause.filter.LogicalExpression;
+import io.stargate.sgv2.jsonapi.exception.ErrorCodeV1;
+import java.util.ArrayList;
+import java.util.List;
+
+public class DBFilterLogicalExpression {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:8]: added, thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+   * This is another enum class for LogicalOperator Different with LogicalOperator in
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:11]: added, thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+    DBLogicalOperator(String operator) {
+      this.operator = operator;
+    }
+
+    public String getOperator() {
+      return operator;
+    }
+
+    public static DBLogicalOperator convertFromLogicalOperator(
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:28]: changed, thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        default ->
+            throw ErrorCodeV1.UNSUPPORTED_FILTER_OPERATION.toApiException(
+                "convert from logical operator failure, unsupported operator: " + logicalOperator);
+      };
+    }
+  }
+
+  private final DBLogicalOperator dbLogicalOperator;
+
+  private final List<DBFilterBase> dbFilterList;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:42]: changed. thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+  private final List<DBFilterLogicalExpression> dbFilterLogicalExpressionList;
+
+  public DBFilterLogicalExpression(DBLogicalOperator dbLogicalOperator) {
+    this.dbLogicalOperator = dbLogicalOperator;
+    this.dbFilterList = new ArrayList<>();
+    this.dbFilterLogicalExpressionList = new ArrayList<>();
+  }
+
+  /**
+   * add a logical relation of DBFilterLogicalExpression to the dbFilterLogicalExpressionList
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:53]: thanks!

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    return innerDBFilterLogicalExpression;
+  }
+
+  /**
+   * add a DBFilterBase to the dbFilterList
+   *
+   * @param DBFilterBase innerDBFilter
+   * @return DBFilterBase
+   */
+  public DBFilterBase addInnerDBFilter(DBFilterBase innerDBFilter) {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:70]: changed.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    dbFilterList.add(innerDBFilter);
+    return innerDBFilter;
+  }
+
+  public DBLogicalOperator getDbLogicalOperator() {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:75]: solved.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+    return dbLogicalOperator;
+  }
+
+  public boolean isEmpty() {
+    return dbFilterLogicalExpressionList.isEmpty() && dbFilterList.isEmpty();
+  }
+
+  public String toString() {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:83]: solved. thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+    StringBuilder sb = new StringBuilder();
+    sb.append("DBFilterLogicalExpression{");
+    sb.append("dbLogicalOperator='").append(dbLogicalOperator).append("'");
+    sb.append(", dbFilterLogicalExpressionList=").append(dbFilterLogicalExpressionList);
+    sb.append(", dbFilterList=").append(dbFilterList);
+    sb.append("}");
+    return sb.toString();
+  }
+
+  public List<DBFilterBase> getDbFilterList() {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:93]: solved.thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+
+  public List<DBFilterLogicalExpression> getDbFilterLogicalExpressionList() {
+    return dbFilterLogicalExpressionList;
+  }
+
+  public int getTotalDbFiltersAmount() {
+    return getTotalDbFiltersAmountHelper(this);
+  }
+
+  private int getTotalDbFiltersAmountHelper(DBFilterLogicalExpression dbFilterLogicalExpression) {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/query/DBFilterLogicalExpression.java:105]: solved, thanks

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroup.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * <p>Is identified by the same Marker as the capture, collects the values the capture matched with
+ * and provides a way for the resolver to pull them out to use in commands.
+ *
+ * <p>Values from all of the operations that matched, they will be multiple values when we start
+ * using greedy matching e.g. match all the string comparison operations in {"username": "foo",
+ * "address.street": "bar"} Path in the {@link ComparisonExpression} that was matched, e.g.
+ * "address.street"
+ *
+ * <p>Created by the {@link CaptureGroups} via a builder <DataType> This is Data type of the object,
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroup.java:23]: solved.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroup.java [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+   *
+   * @param consumer
+   */
+  void consumeAllCaptures(Consumer<CaptureExpression<DataType>> consumer) {
+    captures.forEach(
+        (key, operations) -> {
+          operations.forEach(
+              operation ->
+                  consumer.accept(
+                      new CaptureExpression<DataType>(
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroup.java:40]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * <p>This is "result" of running the FilterMatcher, and the value we pass to the resolver function
+ * so it has raw command and all the groups.
+ *
+ * <p>Each Capture you create from {@link FilterMatcher#capture(Object)} with a Marker is available
+ * here as a {@link CaptureGroup} via {@link #getGroup(Object)}.
+ *
+ * <p>Created in the {@link FilterMatcher} via a builder
+ *
+ * <p>T - The {@link Command} that is filtered against
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:19]: solved.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+ * LogicalExpression, so CaptureGroups needs to have a recursive defined structure
+ */
+public class CaptureGroups<T extends Command> {
+
+  private DBFilterLogicalExpression.DBLogicalOperator dbLogicalOperator;
+  private final T command;
+
+  public final Map<Object, CaptureGroup<?>> captureGroupMap;
+
+  private final List<CaptureGroups<T>> captureGroupsList;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:31]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+ * <p>Created in the {@link FilterMatcher} via a builder
+ *
+ * <p>T - The {@link Command} that is filtered against
+ *
+ * <p>Since we need to keep the logical relation information when doing captures for
+ * LogicalExpression, so CaptureGroups needs to have a recursive defined structure
+ */
+public class CaptureGroups<T extends Command> {
+
+  private DBFilterLogicalExpression.DBLogicalOperator dbLogicalOperator;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:26]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  /**
+   * add next logical relation level of captureGroups to the captureGroupsList this method will be
+   * called when FilterMatcher convert LogicalExpression and populate the captureGroups
+   *
+   * @param captureGroups captureGroups
+   * @return captureGroups
+   */
+  public CaptureGroups<T> addCaptureGroups(CaptureGroups<T> captureGroups) {
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:79]: solved.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+   * consume captures in filterResolver. And since they are in the same logical relation context, so
+   * order does not matter regarding how DB query work.
+   *
+   * <p>
+   *
+   * @param currentDbFilterLogicalExpression, the current DBFilterLogicalExpression to be populated
+   * @param consumer, consumer defines how to consume the current CaptureGroups and current
+   *     DBFilterLogicalExpression.
+   */
+  public void recursiveConsume(
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:115]: changed.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 30]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  /**
+   * The recursive method to help consume captureGroup and convert captures to DBFilters. This
+   * method will be called by resolver method in {@link CollectionFilterResolver} {@link
+   * TableFilterResolver} Resolver method defines the consumer of how to resolve CaptureGroup to
+   * DBFilter. This recursive method will help to consume in a recursive way and populate the
+   * DBFilterLogicalExpression
+   *
+   * <p>The BiConsumer takes two parameters, captureGroups and dBFilterLogicalExpression. it defines
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:91]: moved.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java [hunk 31]
```diff
@@ -29,19 +28,21 @@ public enum MatchStrategy {
 
   private final MatchStrategy strategy;
 
-  private final Function<CaptureExpression, List<DBFilterBase>> resolveFunction;
+  //  private final Function<CaptureExpression, List<DBFilterBase>> resolveFunction;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java:31]: solved.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 32]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * <p>T - The {@link Command} that is filtered against
+ *
+ * <p>Since we need to keep the logical relation information when doing captures for
+ * LogicalExpression, so CaptureGroups needs to have a recursive defined structure
+ */
+public class CaptureGroups<T extends Command> {
+
+  private DBFilterLogicalExpression.DBLogicalOperator dbLogicalOperator;
+  private final T command;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:27]: This is what we have originally, I notice that, think it exists for further purpose?

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java (hunk 2 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java:69]: changed

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java (hunk 3 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java:148]: fixed.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatchRule.java (hunk 4 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatchRule.java:21]: removed.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterResolver.java (hunk 5 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterResolver.java:38]: added back. Though it is unnecessary, since we have matchRules for collection and table.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java (hunk 7 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java:139]: solved.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java (hunk 9 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java:172]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java (hunk 8 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CommandResolver.java:166]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 33]
```diff
@@ -0,0 +1,129 @@
...
+   *
+   * @param currentDbFilterLogicalExpression, the current DBFilterLogicalExpression to be populated
+   * @param consumer, consumer defines how to consume the current CaptureGroups and current
+   *     DBFilterLogicalExpression.
+   */
+  public void recursiveConsume(
+      DBFilterLogicalExpression currentDbFilterLogicalExpression,
+      BiConsumer<CaptureGroups, DBFilterLogicalExpression> consumer) {
+
+    final List<CaptureGroups<T>> captureGroupsList = this.captureGroupsList;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:119]: removed.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java (hunk 3 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/FilterMatcher.java:148]: Discuss

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java [hunk 34]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * <p>Since we need to keep the logical relation information when doing captures for
+ * LogicalExpression, so CaptureGroups needs to have a recursive defined structure
+ */
+public class CaptureGroups<T extends Command> {
+
+  private DBFilterLogicalExpression.DBLogicalOperator dbLogicalOperator;
+  private final T command;
+
+  public final Map<Object, CaptureGroup<?>> captureGroupMap;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java]: solved

**Review by amorton — APPROVED** [review]
approved with some small naming changes

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroup.java [hunk 35]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * update capture entry within the captureGroup for value update, need to merge captures into
+   * single list This is to handle multiple filters with same path within logical operator e.g. if
+   * no merge, then value overrides will happen "$or": [ { "username": "user1" }, { "username":
+   * "user2" } ]
+   *
+   * @param path
+   * @param capture
+   */
+  public void withCapture(String path, List<FilterOperation<DataType>> capture) {
```
[the remark this answers is not in the document]
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroup.java:56]: Good rule, I think, easier to immediately know mutable/non-mutable operations.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java (hunk 6 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java:50]: this is not a quick fix, cause lots of noise in the class.
Let me create a ticket, and do a followup. This has no functional impact.
